smac/utils/validate.py

```diff
@@ -0,0 +1,371 @@
+                config = runhistory.ids_config[key.config_id]
+                if inst in inst_seed_config:
+                    if seed in inst_seed_config[inst]:
+                        inst_seed_config[inst][seed].append(config)
+                    else:
+                        inst_seed_config[inst][seed] = [config]
+                else:
+                    inst_seed_config[inst] = {seed : [config]}
+
+            inst_seed_config = {i :
+                                sorted([(seed, tuple(inst_seed_config[i][seed]))
+                                        for seed in inst_seed_config[i]],
+                                       key=lambda x: len(x[1]))
+                                for i in inst_seed_config}
+        return inst_seed_config
+
+
+    def _get_configs(self, mode):
+        """
+        Return desired configs
+
+        Parameters
+        ----------
+        mode : string
+            from [def, inc, def+inc, time, all], time means eval at 2^0, 2^1, 2^2, ...
+
+        Returns
+        -------
+        configs: list<Configuration>
+            list with desired configurations
+        """
+        # Add desired configs
+        configs = []
+        mode = mode.lower()
+        if mode not in ['def', 'inc', 'def+inc', 'time', 'all']:
+            raise ValueError("%s not a valid option for config_mode in validation."
+                             %mode)
+        if mode == "def" or mode == "def+inc":
+            configs.append(self.scen.cs.get_default_configuration())
+        if mode == "inc" or mode == "def+inc":
+            configs.append(self.traj[-1]["incumbent"])
+        if mode == "time":
+            # add configs at evaluations 2^1, 2^2, 2^3, ...
+            counter = 1
+            for entry in self.traj[:-1]:
+                if (entry["evaluations"] >= counter and
+                        entry["incumbent"] not in configs):
+                    configs.append(entry["incumbent"])
+                    counter *= 2
+            configs.append(self.traj[-1]["incumbent"])  # add last
+        if mode == "all":
+            for entry in self.traj:
+                if not entry["incumbent"] in configs:
+                    configs.append(entry["incumbent"])
+        self.logger.debug("Gathered %d configurations for mode %s.",
+                len(configs), mode)
+        return configs
+
+    def _get_instances(self, mode):
+        """
+        Get desired instances
+
+        Parameters
+        ----------
+        mode: string
+            what instances to use for validation, from [train, test, train+test]
+
+        Returns
+        -------
+        instances: list<strings>
+            instances to be used
+        """
+        instance_mode = mode.lower()
+        if mode not in ['train', 'test', 'train+test']:
+            raise ValueError("%s not a valid option for instance_mode in validation."
+                             %mode)
+
+        instances = dict()
```

Comment:
because i understood instances to be a `dict` with `instance :instance_specifics`.. i'd correct the docstrings rather than the code. or am i on the wrong track?